File: test/features/bills/bill_fetch_screen_data_test.dart
import 'package:billpayments/core/mock/mock_network.dart';
import 'package:billpayments/features/billers/data/biller_catalog_provider.dart';
import 'package:billpayments/features/bills/ui/bill_fetch_form_provider.dart';
import 'package:billpayments/features/bills/ui/bill_fetch_screen_data.dart';
import 'package:billpayments/features/bills/ui/bill_fetch_screen_data_provider.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';

ProviderContainer instantContainer() {
  final container = ProviderContainer(overrides: [
    mockNetworkProvider
        .overrideWithValue(MockNetwork(minDelayMs: 0, maxDelayMs: 1)),
  ]);
  addTearDown(container.dispose);
  return container;
}

void main() {
  test('presentment biller submit enables once all fields filled', () async {
    final container = instantContainer();
    final provider = billFetchScreenDataProvider('electricity-metro');
    final sub = container.listen(provider, (_, _) {});
    await container.read(billerCatalogProvider.future);
    await container.pump();

    var form = container.read(provider) as BillFetchFormData;
    expect(form.billerName, 'Metro Electricity');
    expect(form.inputs.showAmount, isFalse);
    expect(form.submit.action, FetchSubmitAction.fetchBill);
    expect(form.submit.location, isNull);

    container.read(billFetchFormProvider.notifier).editField('account', 'K123');
    form = container.read(provider) as BillFetchFormData;
    expect(
        form.submit.location, '/biller/electricity-metro/review?account=K123');
    sub.close();
  });

  test('typing changes the submit slice but never the inputs slice', () async {
    final container = instantContainer();
    final provider = billFetchScreenDataProvider('electricity-metro');
    final sub = container.listen(provider, (_, _) {});
    await container.read(billerCatalogProvider.future);
    await container.pump();

    final before = (container.read(provider) as BillFetchFormData).inputs;
    container.read(billFetchFormProvider.notifier).editField('account', 'K123');
    final after = container.read(provider) as BillFetchFormData;
    expect(after.inputs, before);
    expect(after.submit.location, isNotNull);
    sub.close();
  });

  test('openAmount biller requires a positive amount', () async {
    final container = instantContainer();
    final provider = billFetchScreenDataProvider('dth-metro');
    final sub = container.listen(provider, (_, _) {});
    await container.read(billerCatalogProvider.future);
    await container.pump();

    final notifier = container.read(billFetchFormProvider.notifier);
    notifier.editField('account', 'D77');
    var form = container.read(provider) as BillFetchFormData;
    expect(form.inputs.showAmount, isTrue);
    expect(form.submit.action, FetchSubmitAction.continueToReview);
    expect(form.submit.location, isNull);

    notifier.editAmount('250');
    form = container.read(provider) as BillFetchFormData;
    expect(form.submit.location,
        '/biller/dth-metro/review?account=D77&amount=25000');
    sub.close();
  });

  test('credit card joins two field values into the account', () async {
    final container = instantContainer();
    final provider = billFetchScreenDataProvider('credit-card-city');
    final sub = container.listen(provider, (_, _) {});
    await container.read(billerCatalogProvider.future);
    await container.pump();

    final notifier = container.read(billFetchFormProvider.notifier);
    notifier.editField('card', '4321');
    notifier.editField('mobile', '9876543210');
    final form = container.read(provider) as BillFetchFormData;
    expect(form.inputs.fields.length, 2);
    expect(form.submit.location,
        '/biller/credit-card-city/review?account=${Uri.encodeQueryComponent('4321|9876543210')}');
    sub.close();
  });
}

File: __tests__/features/bills/billFetchScreenData.test.ts
import { projectBillFetch } from '@/features/bills/ui/billFetchScreenData';
import { asyncData } from '@/core/async/async';
import { makeCatalog } from '@/features/billers/data/billerCatalog';
import { emptyForm, settingValue, settingAmount } from '@/features/bills/ui/billFetchForm';

const catalog = asyncData(makeCatalog([{ id: 'electricity', name: 'Electricity' }, { id: 'dth', name: 'DTH' }, { id: 'credit-card', name: 'Credit Card' }], [
  { id: 'electricity-metro', categoryId: 'electricity', name: 'Metro Electricity', mode: 'presentment', inputParams: [{ key: 'account', label: 'Consumer Number', hint: 'Enter Consumer Number' }] },
  { id: 'dth-metro', categoryId: 'dth', name: 'Metro DTH', mode: 'openAmount', inputParams: [{ key: 'account', label: 'Subscriber ID', hint: 'Enter Subscriber ID' }] },
  { id: 'credit-card-city', categoryId: 'credit-card', name: 'City Credit Card', mode: 'presentment', inputParams: [{ key: 'card', label: 'Card Number', hint: 'Last 4 digits' }, { key: 'mobile', label: 'Registered Mobile', hint: '10-digit mobile' }] },
]));

describe('projectBillFetch', () => {
  it('presentment enables fetchBill once the field is filled', () => {
    const empty = projectBillFetch(catalog, 'electricity-metro', emptyForm());
    expect(empty.kind === 'form' && empty.submit.action).toBe('fetchBill');
    expect(empty.kind === 'form' && empty.submit.location).toBeNull();
    const filled = projectBillFetch(catalog, 'electricity-metro', settingValue(emptyForm(), 'account', 'K123'));
    expect(filled.kind === 'form' && filled.submit.location).toBe('/biller/electricity-metro/review?account=K123');
  });
  it('openAmount requires a positive amount, action continueToReview', () => {
    const f1 = settingValue(emptyForm(), 'account', 'D77');
    const noAmount = projectBillFetch(catalog, 'dth-metro', f1);
    expect(noAmount.kind === 'form' && noAmount.inputs.showAmount).toBe(true);
    expect(noAmount.kind === 'form' && noAmount.submit.action).toBe('continueToReview');
    expect(noAmount.kind === 'form' && noAmount.submit.location).toBeNull();
    const withAmount = projectBillFetch(catalog, 'dth-metro', settingAmount(f1, '250'));
    expect(withAmount.kind === 'form' && withAmount.submit.location).toBe('/biller/dth-metro/review?account=D77&amount=25000');
  });
  it('credit card joins two values with a pipe', () => {
    let form = settingValue(emptyForm(), 'card', '4321');
    form = settingValue(form, 'mobile', '9876543210');
    const result = projectBillFetch(catalog, 'credit-card-city', form);
    expect(result.kind === 'form' && result.submit.location).toBe(`/biller/credit-card-city/review?account=${encodeURIComponent('4321|9876543210')}`);
  });
});
